### PR TITLE
fix: Add `{id}` replacement for ERC-1155

### PR DESCRIPTION
**Pull request by uniyj**
For issue #3010

### CODE REVIEW

**Comment by bigint** [comment]
Screenshot 2022-01-05 at 10 48 47 AM

Cool this works 🚀

File: src/hooks/useENSAvatar.ts [hunk 1]
```diff
@@ -134,11 +135,13 @@ function useERC1155Uri(
   const contract = useERC1155Contract(contractAddress)
   const balance = useSingleCallResult(contract, 'balanceOf', accountArgument)
   const uri = useSingleCallResult(contract, 'uri', idArgument)
+  // ERC-1155 allows a generic {id} in the URL, so prepare to replace if relevant
+  const idHex = id ? BigNumber.from(id).toHexString().substring(2) : id
```
Comment (moodysalem) [inline src/hooks/useENSAvatar.ts:139]: i think it has to be padded to 64 characters

Comment (uniyj) [reply src/hooks/useENSAvatar.ts]: Added leading zero pad to 64 hex chars

File: src/hooks/useENSAvatar.ts [hunk 2]
```diff
@@ -134,11 +136,15 @@ function useERC1155Uri(
...
   const balance = useSingleCallResult(contract, 'balanceOf', accountArgument)
   const uri = useSingleCallResult(contract, 'uri', idArgument)
+  /* ERC-1155 allows a generic {id} in the URL, so prepare to replace if relevant,
+     in lowercase hexadecimal (with no 0x prefix) and leading zero padded to 64 hex characters.
+  */
+  const idHex = id ? hexZeroPad(BigNumber.from(id).toHexString(), 32).substring(2) : id
   return useMemo(
     () => ({
-      uri: !enforceOwnership || balance.result?.[0] > 0 ? uri.result?.[0] : undefined,
+      uri: !enforceOwnership || balance.result?.[0] > 0 ? uri.result?.[0].replace('{id}', idHex) : undefined,
```
Comment (moodysalem) [inline src/hooks/useENSAvatar.ts:145]: this should use `replaceAll`
e.g. `'{id}{id}'.replace('{id}','test')` returns `'test{id}'`
also think you have to use the optional for it, `?.replaceAll`

**Review by moodysalem — CHANGES_REQUESTED** [review]
one more nit

File: src/hooks/useENSAvatar.ts [hunk 3]
```diff
@@ -134,11 +136,15 @@ function useERC1155Uri(
   const contract = useERC1155Contract(contractAddress)
   const balance = useSingleCallResult(contract, 'balanceOf', accountArgument)
   const uri = useSingleCallResult(contract, 'uri', idArgument)
+  /* ERC-1155 allows a generic {id} in the URL, so prepare to replace if relevant,
+     in lowercase hexadecimal (with no 0x prefix) and leading zero padded to 64 hex characters.
+  */
```
[the remark this answers is not in the document]
Comment (uniyj) [reply src/hooks/useENSAvatar.ts:141]: Apologies, was just following the comment style that was added to another recent [PR](https://github.com/Uniswap/interface/pull/3040) that I made. Will change this to `//`.

File: src/hooks/useENSAvatar.ts (hunk 2 above)
Comment (uniyj) [reply src/hooks/useENSAvatar.ts:145]: Ok.

**Pull request merged**